Sources/GRPCHTTP2Core/Client/Resolver/NameResolver.swift

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2024, gRPC Authors All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import GRPCCore
+
+/// A name resolver can provide resolved addresses and service configuration values over time.
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public struct NameResolver: Sendable {
+  /// A sequence of name resolution results.
+  ///
+  /// Resolvers may be push or pull based. Resolvers with the ``UpdateMode-swift.enum/push``
+  /// update mode have addresses pushed to them by an external source and you should subscribe
+  /// to changes in addresses by awaiting for new values in a loop.
+  ///
+  /// Resolvers with the ``UpdateMode-swift.enum/pull`` update mode shouldn't be subscribed to,
+  /// instead you should create an iterator and ask for new results as and when necessary.
+  public var names: RPCAsyncSequence<NameResolutionResult>
+
+  /// How ``names`` is updated and should be consumed.
+  public let updateMode: UpdateMode
+
+  public enum UpdateMode: Hashable, Sendable {
+    /// Addresses are pushed to the resolve by an external source.
+    case push
+    /// Addresses are resolved lazily, when the caller asks them to be resolved.
+    case pull
+  }
+
+  /// Create a new name resolver.
+  public init(results: RPCAsyncSequence<NameResolutionResult>, updateMode: UpdateMode) {
+    self.names = results
+    self.updateMode = updateMode
+  }
+}
+
+/// The result of name resolution, a service of endpoints to connection to and the service
```

Comment:
[Nit] I guess you meant something like:
```suggestion
/// The result of name resolution, a list of endpoints to connect to and the service
```

Reply:
I'm not sure what I was thinking here... 🙈